ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobsCleanupRunner.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright 2023 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.frontend;
+
+import com.github.ambry.named.NamedBlobDb;
+import com.github.ambry.named.StaleNamedResult;
+import com.github.ambry.router.Router;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Pull out stale named blob records, delete those blobs, and soft delete the MySQL db rows
+ */
+public class NamedBlobsCleanupRunner implements Runnable {
+  private final Router router;
+  private final NamedBlobDb namedBlobDb;
+  private static final Logger logger = LoggerFactory.getLogger(NamedBlobsCleanupRunner.class);
+
+  public NamedBlobsCleanupRunner(Router router, NamedBlobDb namedBlobDb) {
+    this.router = router;
+    this.namedBlobDb = namedBlobDb;
+  }
+
+  @Override
+  public void run() {
+    logger.info("Named Blobs Cleanup Runner is initiated");
+    try {
+      List<StaleNamedResult> staleResultList = namedBlobDb.pullStaleBlobIds().get();
+      for (StaleNamedResult staleResult : staleResultList) {
+        router.deleteBlob(staleResult.getBlobId(), "ambry-named-blobs-cleanup-runner").get();
+      }
+      namedBlobDb.cleanupStaleData(staleResultList);
+      logger.info("Named Blobs Cleanup Runner is completed for {} stale cases", staleResultList.size());
```

Comment:
running this in a serial order is fine and probably preferred. However, I think we should not just throw out an exception when there is one stale result failed on delete. We should filter the failed result out and keep move on to the cleanup.

That being said, we should probably limit the number of stale results for each query from named blob DB. And only deal with say 10 at a time. If we get 1M stale result from one query and we will be stuck in delete for a long time. And if one of them failed, we will fail the clean up as well.

so we can probably do
```
List<StaleNamedResult> staleResultList = namedBlob.pullStaleBlobIds().get();
List<StaleNamedResult> resultsToCleanup = new ArrayList<>();
List<StaleNamedResult> failedResults = new ArrayList<>();
while(!staleResultList.isEmpty()) {
   resultsToCleanup.clear();
   for (StaleNameResult result: staleResultList) {
        try {
            router.deleteBlob(result.getBlobId(), "ambry-named-blobs-cleanup-runner").get();
            resultsToCleanup.add(result);
        } catch (Exception e) {
            logger.error("")
            failedResults.add(result);
        }
    }
    namedBlobDb.cleaupStaleData(resultsToCleanup);
    staleResultList = namedBlob.pullStaleBlobIds().get();
    staleResultList.removeAll(failedResults);
}
```

Reply:
+1, we should add some constraint to stop the while loop in runPullStaleBlobIds in case there's a lot in one transaction.

Reply:
Updated with below changes:
1. Limit each staleResultList to be at most 1000 (done via the MySqlNamedBlobDb query side)
2. Catch up exception in each deleteBlob request, and remove those cases in the cleanupStaleData call later.
3. Update the scheduler to run every 10 minutes instead of once per day, to spread out the work more evenly.